File: api/schemas/context.py
"""Schémas Pydantic pour le contexte GDD."""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from api.schemas.dialogue import ContextSelection


class CharacterResponse(BaseModel):
    """Réponse contenant un personnage.
    
    Attributes:
        name: Nom du personnage.
        data: Données complètes du personnage (dict JSON).
    """
    name: str = Field(..., description="Nom du personnage")
    data: Dict[str, Any] = Field(..., description="Données complètes du personnage")


class LocationResponse(BaseModel):
    """Réponse contenant un lieu.
    
    Attributes:
        name: Nom du lieu.
        data: Données complètes du lieu (dict JSON).
    """
    name: str = Field(..., description="Nom du lieu")
    data: Dict[str, Any] = Field(..., description="Données complètes du lieu")


class ItemResponse(BaseModel):
    """Réponse contenant un objet.
    
    Attributes:
        name: Nom de l'objet.
        data: Données complètes de l'objet (dict JSON).
    """
    name: str = Field(..., description="Nom de l'objet")
    data: Dict[str, Any] = Field(..., description="Données complètes de l'objet")


class CharacterListResponse(BaseModel):
    """Réponse contenant une liste de personnages.
    
    Attributes:
        characters: Liste des personnages.
        total: Nombre total de personnages.
        page: Numéro de page actuelle (1-indexed, None si pas de pagination).
        page_size: Taille de la page (None si pas de pagination).
        total_pages: Nombre total de pages (None si pas de pagination).
    """
    characters: List[CharacterResponse] = Field(..., description="Liste des personnages")
    total: int = Field(..., description="Nombre total de personnages")
    page: Optional[int] = Field(None, description="Numéro de page actuelle (1-indexed)")
    page_size: Optional[int] = Field(None, description="Taille de la page")
    total_pages: Optional[int] = Field(None, description="Nombre total de pages")


class LocationListResponse(BaseModel):
    """Réponse contenant une liste de lieux.
    
    Attributes:
        locations: Liste des lieux.
        total: Nombre total de lieux.
        page: Numéro de page actuelle (1-indexed, None si pas de pagination).
        page_size: Taille de la page (None si pas de pagination).
        total_pages: Nombre total de pages (None si pas de pagination).
    """
    locations: List[LocationResponse] = Field(..., description="Liste des lieux")
    total: int = Field(..., description="Nombre total de lieux")
    page: Optional[int] = Field(None, description="Numéro de page actuelle (1-indexed)")
    page_size: Optional[int] = Field(None, description="Taille de la page")
    total_pages: Optional[int] = Field(None, description="Nombre total de pages")


class ItemListResponse(BaseModel):
    """Réponse contenant une liste d'objets.
    
    Attributes:
        items: Liste des objets.
        total: Nombre total d'objets.
        page: Numéro de page actuelle (1-indexed, None si pas de pagination).
        page_size: Taille de la page (None si pas de pagination).
        total_pages: Nombre total de pages (None si pas de pagination).
    """
    items: List[ItemResponse] = Field(..., description="Liste des objets")
    total: int = Field(..., description="Nombre total d'objets")
    page: Optional[int] = Field(None, description="Numéro de page actuelle (1-indexed)")
    page_size: Optional[int] = Field(None, description="Taille de la page")
    total_pages: Optional[int] = Field(None, description="Nombre total de pages")


class SpeciesResponse(BaseModel):
    """Réponse contenant une espèce.
    
    Attributes:
        name: Nom de l'espèce.
        data: Données complètes de l'espèce (dict JSON).
    """
    name: str = Field(..., description="Nom de l'espèce")
    data: Dict[str, Any] = Field(..., description="Données complètes de l'espèce")


class SpeciesListResponse(BaseModel):
    """Réponse contenant une liste d'espèces.
    
    Attributes:
        species: Liste des espèces.
        total: Nombre total d'espèces.
        page: Numéro de page actuelle (1-indexed, None si pas de pagination).
        page_size: Taille de la page (None si pas de pagination).
        total_pages: Nombre total de pages (None si pas de pagination).
    """
    species: List[SpeciesResponse] = Field(..., description="Liste des espèces")
    total: int = Field(..., description="Nombre total d'espèces")
    page: Optional[int] = Field(None, description="Numéro de page actuelle (1-indexed)")
    page_size: Optional[int] = Field(None, description="Taille de la page")
    total_pages: Optional[int] = Field(None, description="Nombre total de pages")


class CommunityResponse(BaseModel):
    """Réponse contenant une communauté.
    
    Attributes:
        name: Nom de la communauté.
        data: Données complètes de la communauté (dict JSON).
    """
    name: str = Field(..., description="Nom de la communauté")
    data: Dict[str, Any] = Field(..., description="Données complètes de la communauté")


class CommunityListResponse(BaseModel):
    """Réponse contenant une liste de communautés.
    
    Attributes:
        communities: Liste des communautés.
        total: Nombre total de communautés.
    """
    communities: List[CommunityResponse] = Field(..., description="Liste des communautés")
    total: int = Field(..., description="Nombre total de communautés")


class RegionListResponse(BaseModel):
    """Réponse contenant une liste de régions.
    
    Attributes:
        regions: Liste des noms de régions.
        total: Nombre total de régions.
    """
    regions: List[str] = Field(..., description="Liste des noms de régions")
    total: int = Field(..., description="Nombre total de régions")


class SubLocationListResponse(BaseModel):
    """Réponse contenant une liste de sous-lieux.
    
    Attributes:
        sub_locations: Liste des noms de sous-lieux.
        total: Nombre total de sous-lieux.
        region_name: Nom de la région.
    """
    sub_locations: List[str] = Field(..., description="Liste des noms de sous-lieux")
    total: int = Field(..., description="Nombre total de sous-lieux")
    region_name: str = Field(..., description="Nom de la région")


class LinkedElementsRequest(BaseModel):
    """Requête pour obtenir les éléments liés.
    
    Attributes:
        character_a: Nom du premier personnage (optionnel).
        character_b: Nom du deuxième personnage (optionnel).
        scene_region: Nom de la région de la scène (optionnel).
        sub_location: Nom du sous-lieu (optionnel).
    """
    character_a: Optional[str] = Field(None, description="Nom du premier personnage")
    character_b: Optional[str] = Field(None, description="Nom du deuxième personnage")
    scene_region: Optional[str] = Field(None, description="Nom de la région de la scène")
    sub_location: Optional[str] = Field(None, description="Nom du sous-lieu")


class LinkedElementsResponse(BaseModel):
    """Réponse contenant les éléments liés.
    
    Attributes:
        linked_elements: Liste des noms d'éléments à sélectionner.
        total: Nombre total d'éléments.
    """
    linked_elements: List[str] = Field(..., description="Liste des noms d'éléments à sélectionner")
    total: int = Field(..., description="Nombre total d'éléments")


class BuildContextRequest(BaseModel):
    """Requête pour construire un contexte personnalisé.
    
    Attributes:
        context_selections: Sélections de contexte GDD.
        user_instructions: Instructions utilisateur.
        max_tokens: Nombre maximum de tokens.
        include_dialogue_type: Inclure le type de dialogue dans le contexte.
    """
    context_selections: ContextSelection = Field(..., description="Sélections de contexte GDD")
    user_instructions: str = Field(default="", description="Instructions utilisateur")
    max_tokens: int = Field(default=1500, ge=100, le=50000, description="Nombre maximum de tokens")
    include_dialogue_type: bool = Field(default=False, description="Inclure le type de dialogue")


class BuildContextResponse(BaseModel):
    """Réponse contenant le contexte construit.
    
    Attributes:
        context: Le contexte construit (texte).
        token_count: Nombre de tokens du contexte.
    """
    context: str = Field(..., description="Contexte construit")
    token_count: int = Field(..., description="Nombre de tokens")

File: frontend/src/api/context.ts
/**
 * API client pour le contexte GDD (personnages, lieux, objets).
 */
import apiClient from './client'
import type {
  CharacterResponse,
  CharacterListResponse,
  LocationResponse,
  LocationListResponse,
  ItemResponse,
  ItemListResponse,
  SpeciesResponse,
  SpeciesListResponse,
  CommunityResponse,
  CommunityListResponse,
  RegionListResponse,
  SubLocationListResponse,
  LinkedElementsRequest,
  LinkedElementsResponse,
} from '../types/api'

/**
 * Liste tous les personnages disponibles.
 */
export async function listCharacters(): Promise<CharacterListResponse> {
  const response = await apiClient.get<CharacterListResponse>('/api/v1/context/characters')
  return response.data
}

/**
 * Récupère un personnage par son nom.
 */
export async function getCharacter(name: string): Promise<CharacterResponse> {
  const response = await apiClient.get<CharacterResponse>(`/api/v1/context/characters/${encodeURIComponent(name)}`)
  return response.data
}

/**
 * Liste tous les lieux disponibles.
 */
export async function listLocations(): Promise<LocationListResponse> {
  const response = await apiClient.get<LocationListResponse>('/api/v1/context/locations')
  return response.data
}

/**
 * Récupère un lieu par son nom.
 */
export async function getLocation(name: string): Promise<LocationResponse> {
  const response = await apiClient.get<LocationResponse>(`/api/v1/context/locations/${encodeURIComponent(name)}`)
  return response.data
}

/**
 * Liste tous les objets disponibles.
 */
export async function listItems(): Promise<ItemListResponse> {
  const response = await apiClient.get<ItemListResponse>('/api/v1/context/items')
  return response.data
}

/**
 * Récupère un objet par son nom.
 * Note: L'endpoint API n'existe pas encore, on filtre depuis la liste.
 */
export async function getItem(name: string): Promise<ItemResponse> {
  const listResponse = await listItems()
  const item = listResponse.items.find((item) => item.name === name)
  if (!item) {
    throw new Error(`Item "${name}" not found`)
  }
  return item
}

/**
 * Liste toutes les espèces disponibles.
 */
export async function listSpecies(): Promise<SpeciesListResponse> {
  const response = await apiClient.get<SpeciesListResponse>('/api/v1/context/species')
  return response.data
}

/**
 * Récupère une espèce par son nom.
 */
export async function getSpecies(name: string): Promise<SpeciesResponse> {
  const response = await apiClient.get<SpeciesResponse>(`/api/v1/context/species/${encodeURIComponent(name)}`)
  return response.data
}

/**
 * Liste toutes les communautés disponibles.
 */
export async function listCommunities(): Promise<CommunityListResponse> {
  const response = await apiClient.get<CommunityListResponse>('/api/v1/context/communities')
  return response.data
}

/**
 * Récupère une communauté par son nom.
 */
export async function getCommunity(name: string): Promise<CommunityResponse> {
  const response = await apiClient.get<CommunityResponse>(`/api/v1/context/communities/${encodeURIComponent(name)}`)
  return response.data
}

/**
 * Liste toutes les régions disponibles.
 */
export async function listRegions(): Promise<RegionListResponse> {
  const response = await apiClient.get<RegionListResponse>('/api/v1/context/locations/regions')
  return response.data
}

/**
 * Récupère les sous-lieux d'une région.
 */
export async function getSubLocations(regionName: string): Promise<SubLocationListResponse> {
  const response = await apiClient.get<SubLocationListResponse>(
    `/api/v1/context/locations/regions/${encodeURIComponent(regionName)}/sub-locations`
  )
  return response.data
}

/**
 * Suggère des éléments liés à partir de personnages et lieux.
 */
export async function getLinkedElements(request: LinkedElementsRequest): Promise<LinkedElementsResponse> {
  const response = await apiClient.post<LinkedElementsResponse>('/api/v1/context/linked-elements', request)
  return response.data
}

